### PR TITLE
`pivot_longer()` with column named `name`

### LINKED ISSUE #692: pivot_longer gives incorrect output when pivoting database column named "name", but gives expected output on local data frame
When a column in a database is called `name`, running `pivot_longer()` on that column results in all of its values being replaced by `"name"` in the `value` column of the long data.  On the other hand, when a column in a local data frame is called `name`, running `pivot_longer()` gives the expected behavior. 

``` r
library(tidyverse)
library(dbplyr)
#> 
#> Attaching package: 'dbplyr'
#> The following objects are masked from 'package:dplyr':
#> 
#>     ident, sql

# Example with a database
db = memdb_frame(name=LETTERS[1:4], id=as.character(1:4), group=rep(c("x","y"), each=2))
db
#> # Source:   table<dbplyr_001> [?? x 3]
#> # Database: sqlite 3.35.5 [:memory:]
#>   name  id    group
#>   <chr> <chr> <chr>
#> 1 A     1     x    
#> 2 B     2     x    
#> 3 C     3     y    
#> 4 D     4     y

# All values in name column get replaced with "name" in the long value column
db %>% pivot_longer(c(name, id))
#> # Source:   lazy query [?? x 3]
#> # Database: sqlite 3.35.5 [:memory:]
#>   group name  value
#>   <chr> <chr> <chr>
#> 1 x     name  name 
#> 2 x     name  name 
#> 3 y     name  name 
#> 4 y     name  name 
#> 5 x     id    1    
#> 6 x     id    2    
#> 7 y     id    3    
#> 8 y     id    4

# Workaround to get desired output
db %>% rename(nm=name) %>% pivot_longer(c(nm, id)) 
#> # Source:   lazy query [?? x 3]
#> # Database: sqlite 3.35.5 [:memory:]
#>   group name  value
#>   <chr> <chr> <chr>
#> 1 x     nm    A    
#> 2 x     nm    B    
#> 3 y     nm    C    
#> 4 y     nm    D    
#> 5 x     id    1    
#> 6 x     id    2    
#> 7 y     id    3    
#> 8 y     id    4

# Another workaround to get desired output
db %>% pivot_longer(c(name, id), names_to="nm")
#> # Source:   lazy query [?? x 3]
#> # Database: sqlite 3.35.5 [:memory:]
#>   group nm    value
#>   <chr> <chr> <chr>
#> 1 x     name  A    
#> 2 x     name  B    
#> 3 y     name  C    
#> 4 y     name  D    
#> 5 x     id    1    
#> 6 x     id    2    
#> 7 y     id    3    
#> 8 y     id    4

# Example with a local data frame
d = tibble(name=LETTERS[1:4], id=as.character(1:4), group=rep(c("x","y"), each=2))
d
#> # A tibble: 4 × 3
#>   name  id    group
#>   <chr> <chr> <chr>
#> 1 A     1     x    
#> 2 B     2     x    
#> 3 C     3     y    
#> 4 D     4     y

# Desired output
d %>% pivot_longer(c(name, id))
#> # A tibble: 8 × 3
#>   group name  value
#>   <chr> <chr> <chr>
#> 1 x     name  A    
#> 2 x     id    1    
#> 3 x     name  B    
#> 4 x     id    2    
#> 5 y     name  C    
#> 6 y     id    3    
#> 7 y     name  D    
#> 8 y     id    4
```

<sup>Created on 2021-08-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### LINKED ISSUE #694: `pivot_longer()` doesn't repair names
`pivot_longer()` doesn't repair names and produces incorrect results

``` r
library(dbplyr)
library(dplyr)
library(tidyr)

df <- memdb_frame(name2 = c("a", "b"), id = 1:2)

df %>% 
  pivot_longer(id, values_to = "name", names_repair = "unique")
#> # Source:   lazy query [?? x 2]
#> # Database: sqlite 3.35.5 [:memory:]
#>   name2  name
#>   <chr> <int>
#> 1 a         1
#> 2 b         2

df %>% 
  collect() %>% 
  pivot_longer(id, values_to = "name", names_repair = "unique")
#> New names:
#> * name -> name...2
#> * name -> name...3
#> # A tibble: 2 × 3
#>   name2 name...2 name...3
#>   <chr> <chr>       <int>
#> 1 a     id              1
#> 2 b     id              2
```

<sup>Created on 2021-08-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### PR DESCRIPTION
**Pull request by mgirlich**
Fixes #692 and fixes #694.

### CODE REVIEW

File: R/verb-pivot-longer.R [hunk 1]
```diff
@@ -147,6 +160,13 @@ dbplyr_pivot_longer_spec <- function(data,
     )
   }
 
+  # TODO remove if clause after `select()` gets early return
+  # https://github.com/tidyverse/dbplyr/pull/691
+  if (any(nms_map$name != nms_map$name_mapped)) {
```
Comment (hadley) [inline R/verb-pivot-longer.R:165]: Can update this now 😄

File: R/verb-pivot-longer.R [hunk 2]
```diff
@@ -110,8 +110,21 @@ dbplyr_pivot_longer_spec <- function(data,
...
+  repair_info <- apply_name_repair_pivot_longer(id_cols, spec, names_repair)
+  id_cols <- repair_info$id_cols
+  spec <- repair_info$spec
 
   spec_split <- vctrs::vec_split(spec, spec[, -(1:2)])
+  nms_map <- tibble(
+    name = colnames(spec_split$key),
+    name_mapped = ifelse(
+      name %in% unlist(spec_split$key),
+      paste0("..", name),
```
Comment (hadley) [inline R/verb-pivot-longer.R:122]: Why add `..` here?

File: R/verb-pivot-longer.R [hunk 3]
```diff
@@ -169,6 +189,26 @@ get_measure_column_exprs <- function(name, value, values_transform) {
   )
 }
 
+apply_name_repair_pivot_longer <- function(id_cols, spec, names_repair) {
+  nms_map_df <- vctrs::vec_rbind(
```
Comment (hadley) [inline R/verb-pivot-longer.R:193]: Would you mind adding a brief comment describing what's happening here?

File: R/verb-pivot-longer.R (hunk 2 above)
Comment (mgirlich) [reply R/verb-pivot-longer.R:122]: I just needed a temporary column name that does not yet exist, so I just prefixed the existing column name with `..`. Of course this column could already exist but this seemed unlikely enough to just do it like this

File: R/verb-pivot-longer.R (hunk 3 above)
Comment (mgirlich) [reply R/verb-pivot-longer.R:193]: Done

File: NEWS.md [hunk 4]
```diff
@@ -30,6 +30,9 @@
 
 * `if_else()` now supports the `missing` argument (@mgirlich, #641).
 
+* `pivot_longer()` can now pivot a column named `name` (@mgirlich, #692).
+
+* `pivot_longer()` can now repair names (@mgirlich, #694).
```
Comment (hadley) [inline NEWS.md:35]: Missing a new line (probably from the merge/rebase)

File: R/verb-pivot-longer.R (hunk 2 above)
Comment (hadley) [reply R/verb-pivot-longer.R:122]: Ok, makes sense.

**Review by hadley — APPROVED** [review]

**Pull request merged**